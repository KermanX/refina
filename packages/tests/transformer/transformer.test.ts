// @refina-ignore

import {
  RefinaTransformer,
  MagicString,
  searchCtxFuncCalls,
} from "@refina/transformer";

describe("transformer", () => {
  it("should transform new file correctly", () => {
    const transformer = new RefinaTransformer();
    const source = `
      _.div(1, _ => {
        _.div(2, _ => {
          _.div(3);
        });
      });
      _.t\`123\`;
      _.t(123);
      _(comp)(1, _ => {})
      _.$cls\`abc\`;
    `;
    const s = new MagicString(source);
    const lastIndex = transformer.transformNewFile(
      s,
      searchCtxFuncCalls(source),
      "~",
    );

    expect(s.toString()).toMatchSnapshot();
    expect(lastIndex).toBe(6);
  });

  it("should return null if nothing is transformed", () => {
    const transformer = new RefinaTransformer();
    const source = `
      123456
    `;
    expect(transformer.transformFile("~", source, false)).toBeNull();
  });

  it("should update transformed file correctly (same)", () => {
    const transformer = new RefinaTransformer();
    const before = `
      _.div(1);
      _.div(2);
    `;
    const after = `
      _.div(1);
      _.div(3);
    `;

    expect(
      transformer.transformFile("~", before, false)?.code,
    ).toMatchSnapshot();
    expect(
      transformer.transformFile("~", after, false)?.code,
    ).toMatchSnapshot();
  });

  it("should update transformed file correctly (append)", () => {
    const transformer = new RefinaTransformer();
    const before = `
      _.div(1);
      _.div(2);
    `;
    const after = `
      _.div(1);
      _.div(2);
      _.div(3);
    `;

    expect(
      transformer.transformFile("~", before, false)?.code,
    ).toMatchSnapshot();
    expect(
      transformer.transformFile("~", after, false)?.code,
    ).toMatchSnapshot();
  });

  it("should update transformed file correctly (insert)", () => {
    const transformer = new RefinaTransformer();
    const before = `
      _.div(1);
      _.span(2);
      _.t\`123\`;
    `;
    const after = `
      _.div(1);
      _.div(2);
      _.span(3);
      _.t\`123\`;
    `;

    expect(
      transformer.transformFile("~", before, false)?.code,
    ).toMatchSnapshot();
    expect(
      transformer.transformFile("~", after, false)?.code,
    ).toMatchSnapshot();
  });

  it("should update transformed file correctly (modify)", () => {
    const transformer = new RefinaTransformer();
    const before = `
      _.div(1);
      _.span(2);
      _.a(2);
    `;
    const after = `
      _.div(1);
      _.div(2);
      _.a(2);
    `;

    expect(
      transformer.transformFile("~", before, false)?.code,
    ).toMatchSnapshot();
    expect(
      transformer.transformFile("~", after, false)?.code,
    ).toMatchSnapshot();
  });

  it("should update transformed file correctly (modify last)", () => {
    const transformer = new RefinaTransformer();
    const before = `
      _.div(1);
      _.span(2);
    `;
    const after = `
      _.div(1);
      _.a(2);
    `;

    expect(
      transformer.transformFile("~", before, false)?.code,
    ).toMatchSnapshot();
    expect(
      transformer.transformFile("~", after, false)?.code,
    ).toMatchSnapshot();
  });

  it("should update transformed file correctly (swap)", () => {
    const transformer = new RefinaTransformer();
    const before = `
      _.div(1);
      _.span(2);
      _.a(2);
      _.div(5);
    `;
    const after = `
      _.div(1);
      _.a(2);
      _.span(2);
      _.div(5);
    `;

    expect(
      transformer.transformFile("~", before, false)?.code,
    ).toMatchSnapshot();
    expect(
      transformer.transformFile("~", after, false)?.code,
    ).toMatchSnapshot();
  });
});
