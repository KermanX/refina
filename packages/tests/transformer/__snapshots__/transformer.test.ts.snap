// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformer > should transform new file correctly 1`] = `
"
      _.$$c(\\"~-1\\", \\"div\\",1, _ => {
        _.$$c(\\"~-2\\", \\"div\\",2, _ => {
          _.$$c(\\"~-3\\", \\"div\\",3);
        });
      });
      _.$$t(\\"~-4\\", \`123\`);
      _.$$t(\\"~-5\\",123);
      _.$$d(\\"~-6\\", comp)(1, _ => {})
      _.$cls\`abc\`;
    "
`;

exports[`transformer > should update transformed file correctly (append) 1`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"div\\",2);
    "
`;

exports[`transformer > should update transformed file correctly (append) 2`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"div\\",2);
      _.$$c(\\"0-3\\", \\"div\\",3);
    "
`;

exports[`transformer > should update transformed file correctly (insert) 1`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"span\\",2);
      _.$$t(\\"0-3\\", \`123\`);
    "
`;

exports[`transformer > should update transformed file correctly (insert) 2`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-4\\", \\"div\\",2);
      _.$$c(\\"0-2\\", \\"span\\",3);
      _.$$t(\\"0-3\\", \`123\`);
    "
`;

exports[`transformer > should update transformed file correctly (modify last) 1`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"span\\",2);
    "
`;

exports[`transformer > should update transformed file correctly (modify last) 2`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-3\\", \\"a\\",2);
    "
`;

exports[`transformer > should update transformed file correctly (modify) 1`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"span\\",2);
      _.$$c(\\"0-3\\", \\"a\\",2);
    "
`;

exports[`transformer > should update transformed file correctly (modify) 2`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-4\\", \\"div\\",2);
      _.$$c(\\"0-3\\", \\"a\\",2);
    "
`;

exports[`transformer > should update transformed file correctly (same) 1`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"div\\",2);
    "
`;

exports[`transformer > should update transformed file correctly (same) 2`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"div\\",3);
    "
`;

exports[`transformer > should update transformed file correctly (swap) 1`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-2\\", \\"span\\",2);
      _.$$c(\\"0-3\\", \\"a\\",2);
      _.$$c(\\"0-4\\", \\"div\\",5);
    "
`;

exports[`transformer > should update transformed file correctly (swap) 2`] = `
"
      _.$$c(\\"0-1\\", \\"div\\",1);
      _.$$c(\\"0-5\\", \\"a\\",2);
      _.$$c(\\"0-6\\", \\"span\\",2);
      _.$$c(\\"0-4\\", \\"div\\",5);
    "
`;
